oracle.go

```diff
@@ -0,0 +1,336 @@
+	}
+	foundIdHex := make(map[string]bool, len(objectIds))
+	for iter.Next(&txnDoc) {
+		foundIdHex[txnDoc.Id.Hex()] = true
+	}
+	if err := iter.Close(); err != nil {
+		if err != mgo.ErrNotFound {
+			// Not found is ok, the transactions may not be complete
+			return nil, err
+		}
+	}
+	result := make(map[string]bool, len(foundIdHex))
+	// because multiple tokens could map to a single txn, we iterate the
+	// passed in tokens instead of caching them in the map.
+	for _, token := range tokens {
+		objIdHex := txnTokenToId(token).Hex()
+		if foundIdHex[objIdHex] {
+			result[token] = true
+		}
+	}
+	o.checkedTokens += uint64(len(tokens))
+	o.completedTokens += uint64(len(result))
+	o.foundTxns += uint64(len(foundIdHex))
+	return result, nil
+}
+
+// RemoveTxns can be used to flag that a given transaction should not
+// be considered part of the valid set.
+func (o *DBOracle) RemoveTxns(txnIds []bson.ObjectId) error {
+	_, err := o.working.RemoveAll(bson.M{"_id": bson.M{"$in": txnIds}})
+	if err != nil {
+		return fmt.Errorf("error removing transaction ids: %v", err)
+	}
+	return nil
+}
+
+type dbIterWrapper struct {
+	iter *mgo.Iter
+}
+
+var _ OracleIterator = (*dbIterWrapper)(nil)
+
+var EOF = fmt.Errorf("end of transaction ids")
+
+func (d *dbIterWrapper) Next() (bson.ObjectId, error) {
+	var txnId struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	if d.iter.Next(&txnId) {
+		return txnId.Id, nil
+	}
+	if err := d.iter.Close(); err != nil {
+		return txnId.Id, err
+	}
+	return txnId.Id, EOF
+}
+func (d *dbIterWrapper) Close() error {
+	return d.iter.Close()
+}
+
+// IterTxns lets you iterate over all of the transactions that have
+// not been removed.
+func (o *DBOracle) IterTxns() (OracleIterator, error) {
+	iter := o.working.Find(nil).Select(bson.M{"_id": 1}).Iter()
+	return &dbIterWrapper{iter: iter}, nil
+}
+
+// MemOracle uses a temporary table on disk to track what transactions are
+// considered completed and purgeable.
+type MemOracle struct {
+	txns            *mgo.Collection
+	completed       map[bson.ObjectId]bool
+	checkedTokens   uint64
+	completedTokens uint64
+	foundTxns       uint64
+}
+
+// NewMemOracle uses an in-memory map to manage the queue of  remaining
+// transactions.
+func NewMemOracle(txns *mgo.Collection) *MemOracle {
+	return &MemOracle{
+		txns: txns,
+	}
+}
+
+var _ Oracle = (*MemOracle)(nil)
+
+func (o *MemOracle) Prepare() (func(), error) {
+	if o.completed != nil {
+		return noopCleanup, fmt.Errorf("Prepare called twice")
+	}
+	// Load the ids of all completed and aborted txns into a separate
+	// temporary collection.
+	logger.Debugf("loading all completed transactions")
+	pipe := o.txns.Pipe([]bson.M{
+		// This used to use $in but that's much slower than $gte.
+		{"$match": bson.M{"s": bson.M{"$gte": taborted}}},
+		{"$project": bson.M{"_id": 1}},
+	})
+	pipe.Batch(maxBatchDocs)
+	pipe.AllowDiskUse()
+	var txnId struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	completed := make(map[bson.ObjectId]bool, 1000)
```

Comment:
Is the capacity hint really worthwhile?

Also, as I understand it, a map with values of `struct{}` is more space efficient that a map of `bool` values. That's slightly less convenient to use when looking to see if a value exists but could be worth it in this case.

https://dave.cheney.net/2014/03/25/the-empty-struct